src/settings/toml/tests/deploy_target.rs

```diff
@@ -335,3 +335,540 @@ fn it_errors_on_deploy_target_routes_and_workers_dev_true() {
 
     assert!(manifest.deploy_target(environment).is_err());
 }
+
+// ENVIRONMENT TESTS
+// Top level empty
+#[test]
+fn when_top_level_empty_env_empty() {
+    let script = "top_level_empty_env_empty";
+    let env_name = "test";
+    let env_config = EnvConfig::default();
+
+    let test_toml = WranglerToml::webpack_with_env(script, env_name, env_config);
+    let toml_string = toml::to_string(&test_toml).unwrap();
+    let manifest = Manifest::new_from_string(toml_string).unwrap();
+
+    let actual_deploy_target = manifest.deploy_target(Some(env_name));
+
+    assert!(actual_deploy_target.is_err());
+
+    // if env only includes zone id, error
+    let mut env_config = EnvConfig::default();
+    env_config.zone_id = Some("samplezoneid");
+
+    let test_toml = WranglerToml::webpack_with_env(script, env_name, env_config);
+    let toml_string = toml::to_string(&test_toml).unwrap();
+    let manifest = Manifest::new_from_string(toml_string).unwrap();
+
+    let actual_deploy_target = manifest.deploy_target(Some(env_name));
+
+    assert!(actual_deploy_target.is_err());
+}
+
+#[test]
+fn when_top_level_empty_zoneless_env() {
+    let env_name = "test";
+
+    // TODO: THIS TEST SHOULD PASS
+    // when env.workers_dev = false
+    let env_config = EnvConfig::zoneless(false);
+
+    let script = "top_level_empty_env_empty";
+    let test_toml = WranglerToml::webpack_with_env(script, env_name, env_config);
+    let toml_string = toml::to_string(&test_toml).unwrap();
+    let manifest = Manifest::new_from_string(toml_string).unwrap();
+
+    let actual_deploy_target = manifest.deploy_target(Some(env_name));
+
+    assert!(actual_deploy_target.is_err());
+
+    // when env.workers_dev = true
+    let env_config = EnvConfig::zoneless(true);
+
+    let script = "top_level_empty_env_zoneless_true";
+    let test_toml = WranglerToml::webpack_with_env(script, env_name, env_config);
+    let toml_string = toml::to_string(&test_toml).unwrap();
+    let manifest = Manifest::new_from_string(toml_string).unwrap();
+
+    let actual_deploy_target = manifest.deploy_target(Some(env_name)).unwrap();
+
+    assert_eq!(actual_deploy_target, DeployTarget::Zoneless);
+}
+
+#[test]
+fn when_top_level_empty_zoned_single_route_env() {
+    let env_name = "test";
+    let zone_id = "samplezoneid";
+
+    // when route is empty, error
+    let pattern = "";
+    let env_config = EnvConfig::zoned_single_route(zone_id, pattern);
+
+    let script = "top_level_empty_env_zoned_single_route_empty";
+    let test_toml = WranglerToml::webpack_with_env(script, env_name, env_config);
+    let toml_string = toml::to_string(&test_toml).unwrap();
+    let manifest = Manifest::new_from_string(toml_string).unwrap();
+
+    let actual_deploy_target = manifest.deploy_target(Some(env_name));
+
+    assert!(actual_deploy_target.is_err());
+
+    // when zone id is missing, error
```

Comment:
each of these sub-tests could honestly could happily live as distinct tests

Reply:
yeah i think you're right. i can break them out. the descriptions became like...two friggin lines though.